### ISSUE TITLE
Paste TABs as SPACEs to clipboard

### ISSUE THREAD

**Issue by halmai** [HeidiSQL/HeidiSQL#1285]
**Is your feature request related to a problem? Please describe.**
I like to use TABs when writing SQL queries. But if I paste a query from HeidiSQL to the putty terminal, the TAB characters cause problems.

**Describe the solution you'd like**
I am aware of the `Preferences` -> `SQL` -> `Tabs to spaces` checkbox but this is not what I want. This would convert my TABs to SPACEs as soon as I type TABs. I would't like it. 

I would also like to keep the TABs as TABs even when I put it to the clipboard (Ctrl-C). I need it in order to have them when I paste it back to Heidi.   

The solution I would like to have is a clipboard content which transfers TABs to SPACEs only if I paste it to a non-heidi application. 

I am not sure if it is technically possible to have both binary and textual representation of something on the clipboard but if it is then this is what I think could work:
- Ctrl+C puts a binary representation of the query to the clipboard
- when Ctrl+V is pressed in Heidi then it pastes it back to Heidi with keeping all the characters (TABs, etc.)
- when Ctrl+V is pressed in any other application that needs a text then the binary content is converted to text with a TAB->SPACE conversion.

If this is not possible then I would like to have a new menu item `Copy with Tabs as Spaces` with hotkey `Shift+Ctrl+C` both under the `Edit` menu and in the right-click menu of the SQL editor as well.

**Comment by WesleyM77** [comment]
Clipboard behavior cannot be changed in the way you're describing.

**Comment by halmai** [comment]
> Clipboard behavior cannot be changed in the way you're describing.

In MS Word, you can put a formatted text to the clipboard. For example, "**ls** _-la_". If you paste it back to Word then the formatting is preserved. If you paste it to a terminal window then it gets converted to plain text. 

I think it should work in Heidi as well: keep the content as formatted text inside Heidi and convert to plain text when pasted to a terminal window.

The reason why I'm not sure and just _think_ is that TAB is not a formatted text, it is just a simple character. But the example shows that the clipboard content can be converted to plain text if needed. In worst case, the query should be stored on the clipboard as a formatted text in which a bold space represents TAB.

**Comment by WesleyM77** [comment]
In your example, the formatting is what MS Word keeps. The terminal has no use for bold or italics when you type them in, so it sheds that information. Tabs and spaces are characters, just like 'X' and a question mark. You're asking for different clipboard contents depending on where you paste, which is the same as asking for one to copy from MS Word the text "ls -la" and expecting to paste "cd example" into terminal, but "ls -la" back into MS Word.

**Comment by halmai** [comment]
I know what a TAB character is. :)

I still think it can be done like this.

- When "<TAB>SELECT * FROM t;" is highlighted in the query editor and the user presses Ctrl-C then Heidi currently puts exactly this text to the clipboard. Instead of this, Heidi should put a formatted text to the clipboard in which "<TAB>" is replaced by a bold space.
- When the user presses Ctrl-V in Heidi then Heidi converts the bold space back to a TAB character.
- When the user presses Ctrl-V in putty (terminal) then putty converts the bold space to simple space. Exactly how it does if I copy it from MS Word.

**Comment by ansgarbecker** [comment]
HeidiSQL already copies a HTML version of the query text to clipboard. I just copied your example from HeidiSQL to LibreOffice Writer:
![grafik](https://user-images.githubusercontent.com/7986591/105216116-6da66780-5b52-11eb-9f33-4219869490cf.png)

So there two formats created on the clipboard by HeidiSQL: a) plain text and b) HTML.
When pasting in Heidi's query editor, currently the "normal" text format from clipboard is taken.

Your suggestion would require HeidiSQL some things to do:
The first clipboard format is plain text, and this is the one your terminal uses, as well as many if not most other applications.
This first text format must have spaces converted to tabs.
Now we need the tab-version in another clipboard format, as you want it for pasting back in HeidiSQL. But the second clipboard format is already in use by the above mentioned HTML version - that is also used by editors like MS Word and whatever. Means HeidiSQL has to register a third clipboard format. This third version has to be used when pasting within HeidiSQL.

Quite much effort I'd say, especially as there already is that "Tabs to spaces" option.

**Comment by ansgarbecker** [comment]
One thing I just realize: in a terminal, spaces might be prettier than a tab, but what about text editors, such as Notepad, or other SQL clients? I suppose the user would wonder why he copied a tab and pasting one or more spaces instead.

**Comment by halmai** [comment]
1.
Let's start with the beginning. :)

> in a terminal, spaces might be prettier than a tab

The problem is not that TABs are not "pretty" in terminal. The problem is that they are totally wrong. For example, if I have the following commands in HeidiSQL

```
SELECT COUNT(*) FROM sent_emails;
		SELECT COUNT(*) FROM sent_emails;
```
(the second line starts with two TABs) then pasting it into `psql` is interpreted as pressing the TAB twice, which lists the available commands:

```
mypgdb=# SELECT COUNT(*) FROM sent_emails;
 count
--------
 15
(1 row)

businessclass=#
ABORT                      CREATE                     FETCH                      REASSIGN                   SET
ALTER                      DEALLOCATE                 GRANT                      REFRESH MATERIALIZED VIEW  SHOW
ANALYZE                    DECLARE                    IMPORT                     REINDEX                    START
BEGIN                      DELETE FROM                INSERT                     RELEASE                    TABLE
CHECKPOINT                 DISCARD                    LISTEN                     RESET                      TRUNCATE
CLOSE                      DO                         LOAD                       REVOKE                     UNLISTEN
CLUSTER                    DROP                       LOCK                       ROLLBACK                   UPDATE
COMMENT                    END                        MOVE                       SAVEPOINT                  VACUUM
COMMIT                     EXECUTE                    NOTIFY                     SECURITY LABEL             VALUES
COPY                       EXPLAIN                    PREPARE                    SELECT                     WITH
mypgdb=# SELECT COUNT(*) FROM sent_emails;
 count
--------
 15
(1 row)
```

You can imagine, having a multi-line query with TABs within the lines can produce a huge mass. This is where my original Feature Request comes from. :) 

2.
The `TABs to spaces` checkbox is far away from what I like. It converts TABs to SPACEs _when I type_. While I navigate, I still need to walk through the lot of spaces which I don't like.

3.
I am not sure how clipboard exactly works in Windows but if I understand your words correctly then the "sender" (HeidiSQL) can place arbitrary number of versions of the content and the "receiver" (the terminal app, LibreOffice Writer, etc.) can decide which version they want. Is this correct?

4.
If it is correct then I don't think HeidiSQL should use three versions. It should put a formatted version with TAB characters and the plain text version with spaces. Putty would receive the plain text version (with spaces), LibreOffice Writer and HeidiSQL would receive the formatted version. 

Do you think it could work?

**Comment by ansgarbecker** [comment]
> he "sender" (HeidiSQL) can place arbitrary number of versions of the content and the "receiver" (the terminal app, LibreOffice Writer, etc.) can decide which version they want. Is this correct?

Yes, that's correct. I did not do that yet, but that's my understanding so far. See [RegisterClipboardFormat](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerclipboardformata) for a deeper look.

> It should put a formatted version with TAB characters and the plain text version with spaces. Putty would receive the plain text version (with spaces), ...

You ignore the fact that there's not just putty. Any other text editor will use the modified plain text format as well. That's the reason why I brought in the third format, as that would be ignored by any other application.

Probably Heidi could offer some explicit "Copy terminal friendly" action (or similar), which gets a different shortcut by default, and does any replacement you want. This would leave the normal copy/paste actions untouched.

**Comment by halmai** [comment]
Yep, it makes sense. Your last sentence concerts with the last sentence of my first comment:

>If this is not possible then I would like to have a new menu item `Copy with Tabs as Spaces` with hotkey `Shift+Ctrl+C` both under the `Edit` menu and in the right-click menu of the SQL editor as well.

**Comment by ansgarbecker** [comment]
Ah yes, well it's a long thread already.
I would not to add that to the context menu, as this is a very special action not used by most users. Instead it gets a default shortcut and should only live in the edit main menu, like the other esoteric things as "Invert selection".

**Comment by halmai** [comment]
> I would not to add that to the context menu, as this is a very special action not used by most users. Instead it gets a default shortcut and should only live in the edit main menu, like the other esoteric things as "Invert selection".

OK, I am happy with this solution. Danke sehr, Herr Becker! ;)
